### PR TITLE
 *: support select partition for partition table

### PR DESCRIPTION
**Pull request by crazycs520**
### What problem does this PR solve? 
support `select partition` syntax, see issue: https://github.com/pingcap/tidb/issues/8414
eg:
```sql
select * from t1 partition (p0,p1)
```

Please merge parser PR first: https://github.com/pingcap/parser/pull/164
Then I will update go.mod in this PR.

### What is changed and how it works?
Add `PartitionNames` to `ast.TableName` and `DataSource`.
Check `PartitionNames` valid when build `buildDataSource`.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Side effects

Related changes

### CODE REVIEW

**Comment by crazycs520** [comment]
@tiancaiamao @ciscoxll PTAL

**Comment by crazycs520** [comment]
Ci failed cause of go.mod, after merge parser PR I will fix it.

File: ddl/partition.go [hunk 1]
```diff
@@ -92,6 +92,7 @@ func buildHashPartitionDefinitions(ctx sessionctx.Context, d *ddl, s *ast.Create
 			return errors.Trace(err)
 		}
 		defs[i].ID = pid
+		defs[i].Name = model.NewCIStr(fmt.Sprintf("p%v", i))
```
Comment (ciscoxll) [inline ddl/partition.go]: s/ fmt.Sprintf("p%v", i) / fmt.Sprintf("%v", i)? `Hash Partition` does not care about partition names.

Comment (crazycs520) [reply ddl/partition.go]: @ciscoxll select partition need p0,p1... name in hash partition table.
In mysql:
```sql
mysql > create table th (a int) partition by hash (a) partitions 5;
Query OK, 0 rows affected
Time: 0.014s
mysql > select * from th partition (p0)
+---+
| a |
+---+
```

Comment (ciscoxll) [reply ddl/partition.go]: Ok, I see.

File: planner/core/logical_plan_builder.go [hunk 2]
```diff
@@ -1994,6 +1995,17 @@ func (b *PlanBuilder) buildDataSource(tn *ast.TableName) (LogicalPlan, error) {
 
 	if tableInfo.GetPartitionInfo() != nil {
 		b.optFlag = b.optFlag | flagPartitionProcessor
+		// check partition
+		if len(tn.PartitionNames) != 0 {
+			for _, name := range tn.PartitionNames {
```
[the remark this answers is not in the document]
Comment (crazycs520) [reply planner/core/logical_plan_builder.go]: will be same with `select * from t partition (p0, p0)`, `prune` function in rule_partition_processor.go will make sure only union the distinct partion.

[the remark this answers is not in the document]
Comment (crazycs520) [reply planner/core/logical_plan_builder.go]: Mysql will also return the result without return error.

[the remark this answers is not in the document]
Comment (crazycs520) [reply planner/core/logical_plan_builder.go]: Aleady add test for it.

**Comment by crazycs520** [comment]
@tiancaiamao @ciscoxll PTAL again

File: executor/executor_test.go [hunk 3]
```diff
@@ -3342,21 +3342,62 @@ func (s *testSuite3) TestSelectHashPartitionTable(c *C) {
 	result.Check(testkit.Rows(
 		"TableReader_8 10.00 root data:Selection_7",
 		"└─Selection_7 10.00 cop eq(test.th.a, -2)",
-		"  └─TableScan_6 10000.00 cop table:th, partition:, range:[-inf,+inf], keep order:false, stats:pseudo",
+		"  └─TableScan_6 10000.00 cop table:th, partition:p2, range:[-inf,+inf], keep order:false, stats:pseudo",
 	))
 	// Test select union all partition.
 	result = tk.MustQuery("desc select * from th;")
 	result.Check(testkit.Rows(
```
Comment (zz-jason) [inline executor/executor_test.go]: @crazycs520 Could you move these explain tests to `explain-test`?

File: planner/core/logical_plan_builder.go [hunk 4]
```diff
@@ -1994,6 +1995,17 @@ func (b *PlanBuilder) buildDataSource(tn *ast.TableName) (LogicalPlan, error) {
 
 	if tableInfo.GetPartitionInfo() != nil {
 		b.optFlag = b.optFlag | flagPartitionProcessor
+		// check partition
+		if len(tn.PartitionNames) != 0 {
```
Comment (zz-jason) [inline planner/core/logical_plan_builder.go]: this check can be removed, the following code block can be simplified to:
```go
for _, name := range tn.PartitionNames {
	// find partition by name
}
```

File: planner/core/rule_partition_processor.go [hunk 5]
```diff
@@ -142,3 +150,13 @@ func (s *partitionProcessor) canBePrune(ctx sessionctx.Context, col *expression.
 	}
 	return len(r) == 0, nil
 }
+
+// findByName checks whether object name exists in list.
+func findByName(list []model.CIStr, name string) bool {
```
Comment (zz-jason) [inline planner/core/rule_partition_processor.go]: 1. Bind it to `partitionProcessor`?
2. s/list/partitionNames/ ?
3. s/name/partitionDef/ ?

File: planner/core/rule_partition_processor.go [hunk 6]
```diff
@@ -97,6 +99,12 @@ func (s *partitionProcessor) prune(ds *DataSource) (LogicalPlan, error) {
 				continue
 			}
 		}
+		// This is for `table partition (p0,p1)` syntax, only union the specified partition if has specified partitions.
+		if len(ds.partitionNames) != 0 {
```
Comment (zz-jason) [inline planner/core/rule_partition_processor.go]: this `if` branch can also be removed.

File: planner/core/rule_partition_processor.go (hunk 5 above)
Comment (crazycs520) [reply planner/core/rule_partition_processor.go]: Why bind to `partitionProcessor`?

File: planner/core/rule_partition_processor.go (hunk 6 above)
Comment (crazycs520) [reply planner/core/rule_partition_processor.go]: No, remove this, `select * from pattirion_table` will return 0 row.

File: planner/core/logical_plan_builder.go (hunk 4 above)
Comment (crazycs520) [reply planner/core/logical_plan_builder.go]: done.

File: planner/core/rule_partition_processor.go (hunk 5 above)
Comment (zz-jason) [reply planner/core/rule_partition_processor.go]: If a function is not intend to be called by multiple objects from different structs, I think it's a good practice to bind it to the struct which calls it.

Comment (crazycs520) [reply planner/core/rule_partition_processor.go]: done.

**Pull request merged**